Library/Homebrew/formula.rb

```diff
@@ -775,6 +775,8 @@ def to_hash
           "used_options" => tab.used_options.as_flags,
           "built_as_bottle" => tab.built_bottle,
           "poured_from_bottle" => tab.poured_from_bottle
+          if ENV["HOMEBREW_CXX"]: "CXX_Compiler" => ENV["HOMEBREW_CXX"]
```

Comment:
Something that may be useful to you is that we store the C++ stdlib which was built against in the tab - GCC's libstdc++ or clang's libc++. If that's the reason you're needing to build boost using GCC, that may give you the information you need.